File: mixin-base/src/main/java/net/yakclient/mixin/base/registry/pool/RegistryPool.java
package net.yakclient.mixin.base.registry.pool;

import net.yakclient.mixin.base.internal.loader.PackageTarget;

import java.util.*;

public abstract class RegistryPool<T> {
    protected final Map<Location, PoolQueue<T>> pool;

    public RegistryPool() {
        this.pool = new HashMap<>();
    }

    public abstract Location pool(T type) throws ClassNotFoundException;

    public abstract PackageTarget register(Location location);

    public void registerAll(Location location) {
        if (this.pool.containsKey(location)) this.register(location);
    }

    public void registerAll() {
        for (Location location : this.pool.keySet()) {
            this.register(location);
        }
    }

    static class PoolQueue<T> {
        protected final Queue<PoolNode<T>> queue;

        public PoolQueue() {
            this.queue = new LinkedList<>();
        }

        public PoolQueue<T> add(T type) {
            this.queue.add(new PoolNode<>(type));
            return this;
        }

        public PoolQueue<T> add(T type, UUID proxy) {
            this.queue.add(new ProxiedPoolNode<>(type, proxy));
            return this;
        }

        public static class PoolNode<T> {
            private final T value;

            PoolNode(T value) {
                this.value = value;
            }

            public T getValue() {
                return this.value;
            }
        }

        public static class ProxiedPoolNode<T> extends PoolNode<T> {
            private final UUID proxy;

            ProxiedPoolNode(T value, UUID proxy) {
                super(value);
                this.proxy = proxy;
            }

            public UUID getProxy() {
                return proxy;
            }
        }
    }
}

File: base/src/main/kotlin/net/yakclient/mixin/base/registry/pool/RegistryPool.kt
package net.yakclient.mixin.base.registry.pool

import java.util.*

abstract class RegistryPool<T> {
    protected val pool: MutableMap<Location, PoolQueue<T>>
    init {
        pool = HashMap()
    }

    @Throws(ClassNotFoundException::class)
    abstract fun pool(type: T): Location

    abstract fun register(location: Location)

    fun registerAll(location: Location) {
        if (pool.containsKey(location)) register(location)
    }

    fun registerAll() {
        for (location in pool.keys) {
            register(location)
        }
    }

    class PoolQueue<T> {
        val queue: Queue<PoolNode<T>>
        init {
            queue = LinkedList()
        }

        fun add(type: T): PoolQueue<T> {
            queue.add(PoolNode(type))
            return this
        }

        fun add(type: T, proxy: UUID): PoolQueue<T> {
            queue.add(ProxiedPoolNode(type, proxy))
            return this
        }

        open class PoolNode<T> internal constructor(val value: T)
        class ProxiedPoolNode<T> internal constructor(value: T, val proxy: UUID) : PoolNode<T>(value)
    }
}